ddl/column.go

```diff
@@ -427,13 +427,106 @@ func (d *ddl) onSetDefaultValue(t *meta.Meta, job *model.Job) error {
 func (d *ddl) onModifyColumn(t *meta.Meta, job *model.Job) error {
 	newCol := &model.ColumnInfo{}
 	oldColName := &model.CIStr{}
-	err := job.DecodeArgs(newCol, oldColName)
+	pos := &ast.ColumnPosition{}
+	err := job.DecodeArgs(newCol, oldColName, pos)
 	if err != nil {
 		job.State = model.JobCancelled
 		return errors.Trace(err)
 	}
 
-	return errors.Trace(d.updateColumn(t, job, newCol, oldColName))
+	if pos.Tp == ast.ColumnPositionNone {
+		return errors.Trace(d.updateColumn(t, job, newCol, oldColName))
+	}
+
+	tblInfo, err := t.GetTable(job.SchemaID, job.TableID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	originalState := tblInfo.State
+	switch tblInfo.State {
+	case model.StatePublic:
+		if err = d.modifyColumnPosition(tblInfo, oldColName, pos); err != nil {
+			job.State = model.JobCancelled
+			return errors.Trace(err)
+		}
+		// use delete only to prevent add record based on different column order.
```

Comment:
If insert without explicit order, plan builder will use the order in schema? So if two node have different schema they will result in two different rows, like `a=1 b=2` and `b=1 a=2`. If we don't use any internal state, we will allow the two different write, because we allow two schema version at same time. But this not seems to be a big problem.